lib/puppet_pal.rb

```diff
@@ -0,0 +1,208 @@
+  # @return [Object] what the Puppet Language code_string evaluates to
+  #
+  def self.evaluate_script_string(code_string)
+    Puppet[:tasks] = true
+    Puppet[:code] = code_string
+    main
+  end
+
+  # Evaluates a Puppet Language script (.pp) file.
+  # @param manifest_file [String] a file with Puppet Language source code
+  # @return [Object] what the Puppet Language manifest_file contents evaluates to
+  #
+  def self.evaluate_script_manifest(manifest_file)
+    Puppet[:tasks] = true
+    main(manifest_file)
+  end
+
+  # Runs the given named plan passing arguments by name in a hash.
+  # @param plan_name [String] the name of the plan to run
+  # @param plan_args [Hash] arguments to the plan - a map of plan parameter name to value, defaults to empty hash
+  # @param manifest_file [String] a Puppet Language file to load and evaluate before running the plan, mutually exclusive with code_string
+  # @param code_string [String] a Puppet Language source string to load and evaluate before running the plan, mutually exclusive with manifest_file
+  # @return [Object] returns what the evaluated plan returns
+  #
+  def self.run_plan(plan_name,
+      plan_args:     {},
+      manifest_file: nil,
+      code_string:   nil
+    )
+    if manifest_file && code_string
+      raise ArgumentError, _("Cannot use 'manifest' and 'code_string' at the same time")
+    end
+
+    unless manifest_file.nil? || manifest_file.is_a?(String)
+      raise ArgumentError, _("Expected 'manifest_file' to be a String, got '%{type}") % { :type => manifest_file.class }
+    end
+
+    unless code_string.nil? || code_string.is_a?(String)
+      raise ArgumentError, _("Expected 'code_string' to be a String, got '%{type}") % { :type => manifest_file.class }
+    end
+
+    Puppet[:tasks] = true
+    Puppet[:code] = code_string unless code_string.nil?
+    main(manifest_file) do | compiler |
+      compiler.topscope.call_function('run_plan', [plan_name, plan_args])
+    end
+  end
+
+  # Defines the context in which to perform puppet operations (evaluation, etc)
+  # The code to evaluate in this context is given in a block.
+  #
+  # @param env_name [String] a name to use for the temporary environment - this only shows up in errors
+  # @param modulepath [Array<String>] an array of directory names containing Puppet modules, may be empty, defaults to empty array
+  # @param settings_hash [Hash] a hash of settings - currently not used for anything, defaults to empty hash
+  # @return [Object] returns what the given block returns
+  # @yieldparam [Puppet::Pal] context, a context that responds to Puppet::Pal methods
+  #
+  def self.in_tmp_environment(env_name,
+      modulepath:    [],
+      settings_hash: {}
+    )
+    unless env_name.is_a?(String) && env_name.length > 0
+      raise ArgumentError(_("Puppet Pal: temporary environment name must be a non empty string, got '%{env_name}'") % {:env_name => env_name})
```

Comment:
this and the next two argumenterrors are wrong syntax

Reply:
Fixed.